### PR TITLE
Add support for SMART app launches against secured endpoint of ISS server for Sandbox launched as a SMART app scenario.

### PR DESCRIPTION
**Pull request by zplata**
Currently, if you launch the Sandbox as a SMART application, the Sandbox will retrieve data from a secured endpoint of the FHIR server it launched against. If a SMART application link were to be returned in a CDS Service response, the app would not be able to be launched because the current only method of launching an app from a card is to append `fhirServiceUrl` and `patientId` query parameters to the launch URL. This assumed using an open fhir server endpoint.

To support SMART app launches against a secured HSPC FHIR server endpoint (for now), we need to append a `launch` and `iss` parameter to SMART launch urls. This scenario occurs if the Sandbox is launched as a SMART application against a secured FHIR endpoint, which is the `iss` parameter to append to the launch urls. The Sandbox can `POST` to the `{base-iss-server-url}/_services/smart/Launch` endpoint with the launch URL of the SMART app link and additional parameters, and if a successful response returns, the Sandbox should receive a launch context ID for the `launch` parameter. See more details about this [here](https://healthservices.atlassian.net/wiki/spaces/HSPC/pages/119734296/Registering+a+Launch+Context).

Note: This launch context registration API is a starting point for launching apps against any arbitrary **HSPC** secured FHIR endpoint. The API endpoint path from a base server is hardcoded in the Sandbox for now, but we may possibly refactor in the future to have this endpoint read from the conformance statement.

@kpshek 
@mjhenkes 
@travistcummings
@kolkheang

### CODE REVIEW

**Comment by travistcummings** [comment]
@zplata Concerning the conformance statement, I think you are proposing that the FHIR server that is also the ISS value contain some info in its conformance statement that says what endpoint to register a SMART app launch context, and the value would be relative to the FHIR server or absolute if it is a different server than the FHIR server.  

When I look at the compatibility statement, the rest.operation does provide a way to specify an OperationDefinition that could contain the input and output parameters.  Are you thinking of using this mechanism of FHIR?

Or are we just extending the rest.security.extension.extension with another property for "launch"?

**Comment by zplata** [comment]
@travistcummings Right, and I was imagining more of the latter, where we would define another object under the `rest.security.extension.extension` like you mentioned, for `launch`. This way, other vendors who implement this endpoint can do so in their own manner while still being able to expose the API endpoint in a consistent way. This was just briefly discussed with @kpshek. Thoughts on this approach?

**Comment by travistcummings** [comment]
I can see one main problem with the launch endpoint property being under rest.security.extension.extension.  That problem is if the FHIR server is open (no security), there would not be any rest.security.extension entry in the conformance statement.  

What would be your preference for an open FHIR server to define such launch registration endpoint?  For HSPC, we have all secured endpoints (with the option of an additional open endpoint) so it isn't such a concern to HSPC.

**Comment by kpshek** [comment]
@travistcummings - I was talking to @zplata offline about the need to support this functionality for more than the HSPC FHIR server since the CDS Hooks Sandbox allows developers to use any FHIR server (for instance, Cerner's or Epic's FHIR servers). As such, the CDS Hooks Sandbox needs to be able to 'discover' where this SMART launch service exists.

While HSPC has this SMART launch service deployed off of the FHIR server endpoint, when we implement this in Cerner's Sandbox I do not want to implement this on our FHIR server. Rather, we will have a dedicated microservice for this particular functionality. Since the CDS Hooks Sandbox just has a FHIR base URL, we'll need to discover the SMART launch service using the FHIR server (hence why Conformance/Capabilities will be used).

Regarding the location of SMART launch endpoint in Conformance, what if it were in `rest.extension`?

I told @zplata (again, offline), that I was fine not addressing this additional change as I didn't want to hold up this PR. It is better to support real SMART app launches today against our default FHIR server in the CDS Hooks Sandbox after all. 😄

@zplata - Can you log a new issue to address supporting SMART app launches for any FHIR server that implements this new strategy we're proposing?

**Comment by kpshek** [comment]
@zplata - Can you give me a CDS Service base URL that I can use to test this change? This CDS Service should return a card that has a SMART app link.

**Comment by zplata** [comment]
@kpshek Issue logged [here](https://github.com/cds-hooks/sandbox/issues/73)

Additionally, you can use Nancy Lush's CDS Service endpoint to test the Bilirubin App they return on the patient view. https://cds-hooks.azurewebsites.net

File: src/scripts/components/Cards.js [hunk 1]
```diff
@@ -47,6 +52,53 @@ const Cards = React.createClass({
     });
   },
 
+  retrievePatientContext(launchUrl) {
```
Comment (kpshek) [inline src/scripts/components/Cards.js]: This function isn't retrieving patient context as much as it is retrieving a SMART launch context

File: src/scripts/components/Cards.js [hunk 2]
```diff
@@ -47,6 +52,53 @@ const Cards = React.createClass({
...
+          patient: this.props.context.get('patient')
+        }
+      }
+    })
+  },
+
+  modifySmartLaunchUrls(card) {
+    return card.links.map(l => {
+      if ("smart" === l.type && !CDS_SMART_OBJ.accessToken) {
+        if (l.url.indexOf("?") < 0) l.url += "?"
```
Comment (kpshek) [inline src/scripts/components/Cards.js]: Something is going wrong with this here as when I test this on an unsecured SMART launch, I'm seeing the URL parameters appended three times. Eg:

`
https://assessment-based-treatment-recommendations.azurewebsites.net/launch.html?fhirServiceUrl=https://sb-fhir-dstu2.smarthealthit.org/api/smartdstu2/open&patientId=SMART-1288992&fhirServiceUrl=https://sb-fhir-dstu2.smarthealthit.org/api/smartdstu2/open&patientId=SMART-1288992&fhirServiceUrl=https://sb-fhir-dstu2.smarthealthit.org/api/smartdstu2/open&patientId=SMART-1288992
`

**Comment by kpshek** [comment]
@zplata - Thanks. How are you launching the CDS Hooks Sandbox (specifically, the deployment of this PR) as a SMART app from the HSPC Sandbox? I'm trying to launch it as a SMART app so that I get a Bearer token to make actual SMART launches.

**Comment by zplata** [comment]
@kpshek I believe I have shared access to my HSPC sandbox (zachsandbox) with you a while back. I have an app configured for heroku that we can test against. I've been looking to update the launch/redirect URLs for this specific deployment instance so we can test this. 

Currently running into a few issues with the HSPC site to update the app config, but I'm talking with @travistcummings off zulip to get them checked out and will update here when we can test this deployment from HSPC.

**Comment by zplata** [comment]
@kpshek Updated with addressed fixes. If you go under the HSPC sandbox, "zachsandbox", navigate to "Launch Scenarios" and you can use the Description, "Use me to test the Sandbox as a SMART app" to launch the Sandbox as a smart app, and test the smart app launch feature.

@travistcummings I sent an invite to you for the hspc sandbox I created for this heroku instance to try out. See my steps above.

File: src/scripts/components/Cards.js (hunk 2 above)
Comment (zplata) [reply src/scripts/components/Cards.js]: Addressed in this PR - looks like when we appended the fhirServiceUrl and patientId params, it was mutating state. Made an object clone of a card that we append these params to so now we shouldn't see that error, and state wont be updated unexpectedly.

**Pull request merged**